### PR TITLE
Bluetooth: Mesh: Fix bt_mesh_net_resend()

### LINKED ISSUE #17907: BLE Mesh when resend use GATT bearer
Hello EvenyBody:
 When I use GATT Bearer to control node with phone. The node sends the long packet at the transport layer, and then is split and sent. The mobile phone sends back an ACK response, but the ACK response does not contain all the response packets, so the node will resend the split packet it thinks is lost. The node calls the function ‘’‘bt_mesh_net_resend‘’‘ to send the retransmitted packet, but not through the GATT but only the Adv Bearer.

```C
bt_mesh_adv_send(buf, cb, cb_data); 
```

### PR DESCRIPTION
**Pull request by jhedberg**
PR #18221 was incorrect in the buffer reference count handling and incomplete in that it lacked calling the send callbacks. This PR tries to address this. See the discussion started by @menglingao1997 at the end of #17907 for details.

Fixes #17907

### CODE REVIEW

File: subsys/bluetooth/mesh/net.c [hunk 1]
```diff
@@ -783,7 +783,7 @@ int bt_mesh_net_resend(struct bt_mesh_subnet *sub, struct net_buf *buf,
 
 	if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY) &&
 	    bt_mesh_proxy_relay(&buf->b, dst)) {
-		net_buf_unref(buf);
+		send_cb_finalize(cb, cb_data);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/net.c]: The advertiser calls its callbacks only after the packet is actually sent on air, I think we should do the same for proxy, using bt_gatt_notify_cb. There'll be a race between adv and gatt, which complicates things, but we could get rid of the corner cases that caused this bug by generalizing this.

A different PR, probably, though. I can take a look at it if you agree and don't see any big blockers.

File: subsys/bluetooth/mesh/net.h [hunk 2]
```diff
@@ -362,3 +362,15 @@ struct friend_cred *friend_cred_create(struct bt_mesh_subnet *sub, u16_t addr,
 				       u16_t lpn_counter, u16_t frnd_counter);
 void friend_cred_clear(struct friend_cred *cred);
 int friend_cred_del(u16_t net_idx, u16_t addr);
+
+static inline void send_cb_finalize(const struct bt_mesh_send_cb *cb,
+				    void *cb_data)
+{
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/net.h]: This needs an `if (cb)` check

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/net.c (hunk 1 above)
Comment (jhedberg) [reply subsys/bluetooth/mesh/net.c]: Yeah, I was thinking of adding a "TODO" note in the code not to forget - it'll require some changes to the pipeline of function calls in proxy.c to get the callback information propagated to the point where it gets passed to the GATT layer using bt_gatt_notify_cb

File: subsys/bluetooth/mesh/net.h (hunk 2 above)
Comment (jhedberg) [reply subsys/bluetooth/mesh/net.h]: All current callers use `if (cb) send_cb_finalize()`. I can move it into the function if you think that's cleaner - I don't have a strong opinion on it.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/net.h]: Not anymore? Or perhaps it's still too early in the morning for me

Comment (jhedberg) [reply subsys/bluetooth/mesh/net.h]: No, early morning for me 😄 One place had it, another didn't. I'll just move it to the function.

Comment (jhedberg) [reply subsys/bluetooth/mesh/net.h]: @trond-snekvik should be fixed now

**Review by trond-snekvik — APPROVED** [review]

**Review by joerchan — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**